Separate data_index lookups from parsing

cben opened the pull request:
Openshift half of https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/46, same motivation & approach.

- Images only partially untangled — the real complication stayed in kubernetes `parse_container_image` reading and writing `data_index` for both images & registries :-(
  I probably can live with that for initial graph refresh.

@miq-bot add-label refactoring
@enoodle @zeari @moolitayer @zakiva @yaacov Please review.

- [x] also tested locally against deletion tests #18.

towards RFE: https://bugzilla.redhat.com/show_bug.cgi?id=1470021

cben commented:
master is red, Travis should be fixed after #31

enoodle commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -14,37 +14,66 @@ def ems_inv_to_hashes(inventory, options = Config::Options.new)
-          @data_index.store_path(key, :by_name, ns[:name], ns)
+        @data[key].each do |b|
+          b[:project] = @data_index.fetch_path(path_for_entity("namespace"), :by_name, b[:namespace])
+          # TODO: BuildConfig is namespaced, should index by_namespace_and_name.
```
:+1:

cben commented:
Rebased on top of @enoodle's bug fix #33.

simon3z commented:
@cben can you squash the commits?

cben commented:
I'd rather not squash.  These are paired remove-here-add-there refactorings, each self-contained (and passing tests), while squashed it will be a big confusing soup.

cben commented:
ping @enoodle @zeari @moolitayer @zakiva @yaacov, please review (easier by commits)

simon3z commented:
@moolitayer can review/approve?

moolitayer commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -39,12 +39,16 @@ def get_routes(inventory)
-
-        @data[key].each do |ns|
-          @data_index.store_path(key, :by_name, ns[:name], ns)
+        key = path_for_entity("namespace")
```
`key = path_for_entity("project")` ?
(is it the same?)

cben replied:
it's the same, both return `:container_projects`.
I wanted to emphasize that we're merging into same place that kubernetes parser `get_namespaces` fills...

I'm wondering if renaming this method to say `merge_projects_into_namespaces` would be good.

simon3z replied:
> I'm wondering if renaming this method to say merge_projects_into_namespaces would be good.

@cben :+1:

cben commented:
project `display_name` parsing had no tests, working on test now...

cben commented:
- added display_name test.
- renamed the merging methods.
- deliberately ignoring rubocop on L485, would be hard to read without braces.
- codeclimate irrelevant (complains about whole file following a consistent pattern).

@moolitayer @simon3z ready for merge?

moolitayer commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -3,48 +3,77 @@ module Openshift
       def ems_inv_to_hashes(inventory, options = Config::Options.new)
         super(inventory, options)
-        get_projects(inventory)
+        merge_projects_into_namespaces(inventory)
```
:+1:

moolitayer commented:
> @moolitayer @simon3z ready for merge?

it has my approval, :+1:

cben commented:
@simon3z @moolitayer I'm reviving #9 and will need to backport it to euwe & fine.
The backport will be much cleaner if I can backport the image parts of this refactoring (commits adfc20598698f55decd2beb34d4cf33d966a3d17, e18e092c30d439f678e22655e09b6edeb2dfbb93).
Do you that will be acceptable?  How strict is the "we don't backport refactorings" rule?
